packages/eip1193-permission-middleware/src/wallet-requestPermissions.ts

```diff
@@ -0,0 +1,175 @@
+  methodNames: [MethodNames.RequestPermissions],
+  implementation: requestPermissionsImplementation,
+  hookNames: {
+    getAccounts: true,
+    requestPermissionsForOrigin: true,
+    getCaip25PermissionFromLegacyPermissionsForOrigin: true,
+  },
+};
+
+type AbstractPermissionController = PermissionController<
+  PermissionSpecificationConstraint,
+  CaveatSpecificationConstraint
+>;
+
+type GrantedPermissions = Awaited<
+  ReturnType<AbstractPermissionController['requestPermissions']>
+>[0];
+
+/**
+ * Request Permissions implementation to be used in JsonRpcEngine middleware, specifically for `wallet_requestPermissions` RPC method.
+ * The request object is expected to contain a CAIP-25 endowment permission.
+ *
+ * @param req - The JsonRpcEngine request
+ * @param res - The JsonRpcEngine result object
+ * @param _next - JsonRpcEngine next() callback - unused
+ * @param end - JsonRpcEngine end() callback
+ * @param options - Method hooks passed to the method implementation
+ * @param options.getAccounts - A hook that returns the permitted eth accounts for the origin sorted by lastSelected.
+ * @param options.getCaip25PermissionFromLegacyPermissionsForOrigin - A hook that returns a CAIP-25 permission from a legacy `eth_accounts` and `endowment:permitted-chains` permission.
+ * @param options.requestPermissionsForOrigin - A hook that requests CAIP-25 permissions for the origin.
+ * @returns Nothing.
+ */
+async function requestPermissionsImplementation(
+  req: JsonRpcRequest<[RequestedPermissions]> & { origin: string },
+  res: PendingJsonRpcResponse<Json>,
+  _next: AsyncJsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback,
+  {
+    getAccounts,
+    requestPermissionsForOrigin,
+    getCaip25PermissionFromLegacyPermissionsForOrigin,
+  }: {
+    getAccounts: () => string[];
+    requestPermissionsForOrigin: (
+      requestedPermissions: RequestedPermissions,
+    ) => Promise<[GrantedPermissions]>;
+    getCaip25PermissionFromLegacyPermissionsForOrigin: (
+      requestedPermissions?: RequestedPermissions,
+    ) => RequestedPermissions;
+  },
+) {
+  const { params } = req;
+
+  if (!Array.isArray(params) || !isPlainObject(params[0])) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  let [requestedPermissions] = params;
+  delete requestedPermissions[Caip25EndowmentPermissionName];
+
+  const caip25EquivalentPermissions: Partial<
+    Pick<RequestedPermissions, 'eth_accounts' | 'endowment:permitted-chains'>
+  > = pick(requestedPermissions, [
+    RestrictedMethods.EthAccounts,
+    EndowmentTypes.PermittedChains,
+  ]);
+  delete requestedPermissions[RestrictedMethods.EthAccounts];
+  delete requestedPermissions[EndowmentTypes.PermittedChains];
+
+  const hasCaip25EquivalentPermissions =
+    Object.keys(caip25EquivalentPermissions).length > 0;
+
+  if (hasCaip25EquivalentPermissions) {
+    const caip25Permission = getCaip25PermissionFromLegacyPermissionsForOrigin(
```

Comment:
should we move the `getCaip25PermissionFromLegacyPermissions` hook over from extension? The only platform specific thing I may be worried about is the special handling for snap origins

Reply:
ahh sorry. ignore this. definitely outside the scope of this PR. my bad